src/Libraries/Microsoft.Extensions.AI/ChatCompletion/OpenTelemetryChatClient.cs

```diff
@@ -162,122 +182,97 @@ public override async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteSt
-                }
-
-                _ = sb.Append(']');
+                choices[update.ChoiceIndex] = choiceContents = [];
             }
 
-            _ = sb.Append('}');
-        }
-
-        _ = sb.Append(']');
-        return sb.ToString();
-    }
-
-    /// <summary>Organize streaming content by choice index.</summary>
-    private static Dictionary<int, List<StreamingChatCompletionUpdate>> OrganizeStreamingContent(IEnumerable<StreamingChatCompletionUpdate>? contents)
-    {
-        Dictionary<int, List<StreamingChatCompletionUpdate>> choices = [];
-        if (contents is null)
-        {
-            return choices;
+            choiceContents.Add(update);
         }
 
-        foreach (var content in contents)
+        // Add a ChatMessage for each choice.
+        string? id = null;
+        ChatFinishReason? finishReason = null;
+        string? modelId = null;
+        List<ChatMessage> messages = new(choices.Count);
+        foreach (var choice in choices.OrderBy(c => c.Key))
         {
-            if (!choices.TryGetValue(content.ChoiceIndex, out var choiceContents))
+            ChatRole? role = null;
+            List<AIContent> items = [];
+            foreach (var update in choice.Value)
             {
-                choices[content.ChoiceIndex] = choiceContents = [];
+                id ??= update.CompletionId;
+                finishReason ??= update.FinishReason;
+                role ??= update.Role;
+                items.AddRange(update.Contents);
+                modelId ??= update.Contents.FirstOrDefault(c => c.ModelId is not null)?.ModelId;
             }
 
-            choiceContents.Add(content);
+            messages.Add(new ChatMessage(role ?? ChatRole.Assistant, items));
         }
 
-        return choices;
+        return new(messages)
+        {
+            CompletionId = id,
+            FinishReason = finishReason,
+            ModelId = modelId,
+            Usage = updates.SelectMany(c => c.Contents).OfType<UsageContent>().LastOrDefault()?.Details,
+        };
     }
 
     /// <summary>Creates an activity for a chat completion request, or returns null if not enabled.</summary>
-    private Activity? StartActivity(IList<ChatMessage> chatMessages, ChatOptions? options)
+    private Activity? CreateAndConfigureActivity(ChatOptions? options)
     {
         Activity? activity = null;
-        if (Enabled)
+        if (_activitySource.HasListeners())
         {
             string? modelId = options?.ModelId ?? _modelId;
 
             activity = _activitySource.StartActivity(
-                $"chat.completions {modelId}",
+                $"{OpenTelemetryConsts.GenAI.Chat} {modelId}",
                 ActivityKind.Client,
                 default(ActivityContext),
                 [
-                    new(OpenTelemetryConsts.GenAI.Operation.Name, "chat"),
+                    new(OpenTelemetryConsts.GenAI.Operation.Name, OpenTelemetryConsts.GenAI.Chat),
```

Comment:
>                   new(OpenTelemetryConsts.GenAI.Operation.Name, OpenTelemetryConsts.GenAI.Chat), [](http://example.com/codeflow?start=1&length=97)

maybe you can delay adding the tags after checking if the activity is null? I am not sure if the allocation is sensitive here or not.

Also, we have TagList which can be used to avoid the array allocation?